shared-bindings/displayio/Bitmap.c

```diff
@@ -172,7 +172,106 @@ STATIC mp_obj_t bitmap_subscr(mp_obj_t self_in, mp_obj_t index_obj, mp_obj_t val
     return mp_const_none;
 }
 
-//|     def fill(self, value: int) -> None:
+//|     def blit(self, x: int, y: int, source_bitmap: bitmap, *, x1: int, y1: int, x2: int, y2: int, skip_index: int) -> Any:
+//|         """Inserts the source_bitmap region defined by rectangular boundaries
+//|         (x1,y1) and (x2,y2) into the bitmap at the specified (x,y) location.
+//|         :param int x: Horizontal pixel location in bitmap where source_bitmap upper-left
+//|             corner will be placed
+//|         :param int y: Vertical pixel location in bitmap where source_bitmap upper-left
+//|             corner will be placed
+//|         :param bitmap source_bitmap: Source bitmap that contains the graphical region to be copied
+//|         : param int x1: Minimum x-value for rectangular bounding box to be copied from the source bitmap
+//|         : param int y1: Minimum y-value for rectangular bounding box to be copied from the source bitmap
+//|         : param int x2: Maximum x-value for rectangular bounding box to be copied from the source bitmap
+//|         : param int y2: Maximum y-value for rectangular bounding box to be copied from the source bitmap
+//|         : param int skip_index: bitmap palette index in the source that will not be copied,
+//|             set `None` to copy all pixels"""
+//|         ...
+//|
+
+STATIC mp_obj_t displayio_bitmap_obj_blit(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args){
+    enum {ARG_x, ARG_y, ARG_source, ARG_x1, ARG_y1, ARG_x2, ARG_y2, ARG_skip_index};
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_x, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_y, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_source_bitmap, MP_ARG_REQUIRED | MP_ARG_OBJ},
+        {MP_QSTR_x1, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = 0} },
+        {MP_QSTR_y1, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = 0} },
+        {MP_QSTR_x2, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} }, // None convert to source->width
+        {MP_QSTR_y2, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} }, // None convert to source->height
+        {MP_QSTR_skip_index, MP_ARG_OBJ | MP_ARG_KW_ONLY, {.u_obj=mp_const_none} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    displayio_bitmap_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+
+    int16_t x = args[ARG_x].u_int;
+    int16_t y = args[ARG_y].u_int;
+
+    displayio_bitmap_t *source = MP_OBJ_TO_PTR(args[ARG_source].u_obj);
+
+    // ensure that the target bitmap (self) has at least as many `bits_per_value` as the source
+    if (self->bits_per_value < source->bits_per_value) {
+        mp_raise_ValueError(translate("Cannot blit: source palette too large."));
+    }
+
+    int16_t x1, y1, x2, y2;
```

Comment:
Right now x1 and y1 won't work, they'll always be 0.

```suggestion
    int16_t x1 = args[ARG_x1].u_int;
    int16_t y1 = args[ARG_y1].u_int;
    int16_t x2, y2;
```

Reply:
I updated my test `code.py` to verify the correct operation of `x1` and `y1` variables (see last `blit` command).  

```
import displayio

bmap_group = displayio.Group(max_size=2)  # Create a group for displaying

bitmap1=displayio.Bitmap(50, 50, 4)
bitmap2=displayio.Bitmap(50, 50, 4)
bitmap3=displayio.Bitmap(50, 50, 4)
bitmap4=displayio.Bitmap(10, 10, 4)
bitmap5=displayio.Bitmap(10, 10, 8)
bitmap6=displayio.Bitmap(100,100,4)
palette=displayio.Palette(4)
palette[0]=0x000000
palette[1]=0xFF0000
palette[2]=0x00FF00
palette[3]=0x0000FF

bitmap1.fill(1)
bitmap2.fill(2)
bitmap3.fill(3)
bitmap4.fill(0)

mytilegrid=displayio.TileGrid(bitmap1, pixel_shader=palette, width=1, height=1, 
                            tile_width=bitmap1.width, tile_height=bitmap1.height,
                            default_tile=0,
                            x=10, y=10)

mytilegrid2=displayio.TileGrid(bitmap6, pixel_shader=palette, width=1, height=1, 
                            tile_width=bitmap6.width, tile_height=bitmap6.height,
                            default_tile=0,
                            x=70, y=70)

bitmap_group=displayio.Group(max_size=2)
bitmap_group.append(mytilegrid)
bitmap_group.append(mytilegrid2)
display.auto_refresh = True

display.show(bitmap_group)

time.sleep(2)
bitmap1.blit(10, 10, bitmap2, skip_index=2)  # this should do nothing, since the 2 palette index is not copied
time.sleep(2)
#bitmap1.blit(0, 0, bitmap2, skip_index=None, x1=0, y1=0, x2=49, y2=49) # leave one pixel around lower right corner
#time.sleep(2)
bitmap1.blit(10, 10, bitmap2, skip_index=3)
time.sleep(2)
bitmap1.blit(3, 3, bitmap3, x1=5, y1=5, x2=15, y2=25)
time.sleep(2)
bitmap1.blit(30, 30, bitmap4)
time.sleep(2)
bitmap1.blit(10, 30, bitmap4)
time.sleep(2)
bitmap1.blit(5, 10, bitmap4)
time.sleep(2)
bitmap1.blit(10, 5, bitmap4)
time.sleep(2)
bitmap1.blit(0, 0, bitmap3, x1=25, y1=45, x2=25, y2=45)
print('got here')
bitmap1.blit(0, 0, bitmap3, x1=25, y1=45)

time.sleep(2)
bitmap1.blit(25, 25, bitmap3, x1=25, y1=10, y2=40)
time.sleep(2)

bitmap6.blit(0,0,bitmap1)
time.sleep(0.5)
bitmap6.blit(0,50,bitmap1)
time.sleep(0.5)
bitmap6.blit(50,0,bitmap1)
time.sleep(0.5)
bitmap6.blit(65, 65,bitmap1, x1=15, y1=15)

time.sleep(60)
```